plugins: ensure to log the offending plugin on instantiation failure

wisp3rwind opened the pull request:
Came up in https://github.com/beetbox/beets/pull/5701#discussion_r2049220759 where an error can't easily be traced back to a plugin.

(Draft because I want to avoid creating merge conflicts for https://github.com/beetbox/beets/pull/5701 before that is merged<del>, and because I might a few additional improvements here</del>.)

wisp3rwind committed:
plugins: ensure to log the offending plugin on instantiation failure

wisp3rwind committed:
plugins: flatten loading code

minimize how much code is wrapped in a given try...except statement

wisp3rwind committed:
plugins: catch the "new" ModuleNotFoundError instead of old hacky & broken error handling

from a quick test, it appears that the old code shouldn't even be working
anymore due to additional quotes in exc.args[0]

snejus commented on beets/plugins.py:
```diff
@@ -270,31 +270,28 @@ def load_plugins(names=()):
-                        _classes.add(obj)
-
-        except Exception:
+            namespace = __import__(modname, None, None)
```
Perhaps we could use the more standard `importlib.import_module`?

snejus commented on beets/plugins.py:
```diff
@@ -270,31 +270,28 @@ def load_plugins(names=()):
             )
+            continue
+
+        for obj in getattr(namespace, name).__dict__.values():
```
Consider using `inspect.getmembers`

wisp3rwind replied:
:+1:

snejus commented on beets/plugins.py:
```diff
@@ -270,31 +270,28 @@ def load_plugins(names=()):
+
+        for obj in getattr(namespace, name).__dict__.values():
+            if (
+                isinstance(obj, type)
```
... and `inspect.isclass` here

wisp3rwind replied:
Good call!

snejus commented on beets/plugins.py:
```diff
@@ -315,7 +312,15 @@ def find_plugins():
+                    cls.__module__,
+                    cls.__qualname__,
+                )
+                raise
```
Shouldn't we just allow it to die?

wisp3rwind replied:
Maybe I mis-read this and thought you were commmenting on another line: This re-raise should lead to beets exiting.
However, it won't exit on import failures (a few lines above).

snejus replied:
I was really unclear here in any case 🤣 

I was meant to say: shall we handle plugin failure gracefully? Ideally, a single plugin failing to initialise should not stop beets execution. I'm thinking of something like

```python
def find_plugins() -> Iterable[BeetsPlugin]:
    """Returns a list of BeetsPlugin subclass instances from all
    currently loaded beets plugins. Loads the default plugin set
    first.
    """
    if not _instances:
        load_plugins()
        for _class in list(_classes):
            try:
                _instances[_class] = _class()
            except Exception:
                log.error(...)
                _classes.remove(_class)

    return _instances.values()
```

wisp3rwind replied:
Interesting, so we actually have completely opposite opinions here :laughing:

In any case; I'd say that things are not really consistent now:
- if a plugin is not found (e.g. because it was pip-installed for Python 3.X, but then a Python update to 3.(X+1) happened) or can't be imported (e.g. because there are syntax errors), we just warn about it
- if a plugin can be imported, but raises on `__init__`, we abort/crash

My reasoning is that some plugin actions are actually tricky to recover from if unexpectedly not run during import. For example `copyartifacts` (although that is broken anyway), `fetchart` with local art files that are in the same folder at the to-be-imported album, and probably more. Others are easier to recover from (say fetching lyrics, where we only have remote sources IIRC).

So, this seems to warrant some discussion, see #5740. Let's keep this PR focused on code cleanup/debugging improvements, and not change behavior.

wisp3rwind committed:
plugins: use importlib and inspect instead __import__ and __dict__

also remove a unnecessary check before adding to a set

wisp3rwind committed:
plugins: remove no-op load_plugins call

without an argument (i.e. the default `names=()`), this was a no-op
actual plugin loading is triggered by tests, or by the ui code

-> Turns out it's not an actual no-op in our testing code due to
   patching with attaching some side-effects to load_plugins...
   This is fixed up in the next commit.

wisp3rwind committed:
plugins: actually load plugins in load_plugins()

Instead of the first call to find_plugins(). This could change behavior
slightly, since plugin load (and errors that happen in the process) could
occur earlier now. But:

- I'd argue this is better anyway, since plugin load errors now happen
  in the initial call to load_plugins() instead of somewhere in the
  middle of the first operation that uses plugins.
- In our ui code, both happen immediately after each other anyway:
  ui._setup calls
  	1) ui._load_plugins -> plugins.load_plugins
	2) plugins.types, plugins.named_queries, ...

wisp3rwind commented:
This should be ready to go now; it's easiest to review commit-by-commit.

snejus commented on beets/plugins.py:
```diff
@@ -350,57 +351,55 @@ def load_plugins(names: Sequence[str] = ()) -> None:
-                        _classes.add(obj)
-
-        except Exception:
+            mod = import_module(f".{name}", package=PLUGIN_NAMESPACE)
```
This may be slightly misleading in some cases, for example:

```py
# beetsplug/discogs.py
...
import beets.ibrary
```

```sh
$ beet import -LI sel
...
** plugin discogs not found
```

snejus commented on beets/plugins.py:
```diff
@@ -350,57 +351,55 @@ def load_plugins(names: Sequence[str] = ()) -> None:
-
-        except Exception:
+            mod = import_module(f".{name}", package=PLUGIN_NAMESPACE)
+        except ModuleNotFoundError:
```
Maybe we should catch a generic `Exception` here? Otherwise every other exception will make beets die, for example:

```py
# beetsplug/discogs.py
...
abc
```

```sh
$ beet import -LI sel
...
Traceback (most recent call last):
  File "/home/sarunas/.local/bin/beet", line 5, in <module>
    main()
  File "/home/sarunas/repo/beets/beets/ui/__init__.py", line 1878, in main
    _raw_main(args)
  File "/home/sarunas/repo/beets/beets/ui/__init__.py", line 1853, in _raw_main
    subcommands, plugins, lib = _setup(options, lib)
  File "/home/sarunas/repo/beets/beets/ui/__init__.py", line 1680, in _setup
    plugins = _load_plugins(options, config)
  File "/home/sarunas/repo/beets/beets/ui/__init__.py", line 1669, in _load_plugins
    plugins.load_plugins(plugin_list)
  File "/home/sarunas/repo/beets/beets/plugins.py", line 358, in load_plugins
    mod = import_module(f".{name}", package=PLUGIN_NAMESPACE)
  File "/home/sarunas/.local/share/pyenv/versions/3.9.20/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/home/sarunas/repo/beets/beetsplug/discogs.py", line 32, in <module>
    abc
NameError: name 'abc' is not defined
```

snejus commented on beets/plugins.py:
```diff
@@ -350,57 +351,55 @@ def load_plugins(names: Sequence[str] = ()) -> None:
+        for _name, cls in inspect.getmembers(mod, inspect.isclass):
+            if (
+                issubclass(cls, BeetsPlugin)
+                and cls != BeetsPlugin
+                and cls != MetadataSourcePlugin
```
```suggestion
                and cls not in {BeetsPlugin, MetadataSourcePlugin}
```

wisp3rwind replied:
:+1:

snejus commented on beets/plugins.py:
```diff
@@ -350,57 +351,55 @@ def load_plugins(names: Sequence[str] = ()) -> None:
+                issubclass(cls, BeetsPlugin)
+                and cls != BeetsPlugin
+                and cls != MetadataSourcePlugin
+                and cls not in _instances
```
Do we need to perform this check? It's unlikely we will come across dupes, and even if we do, `classes` is defined as a set, so no dupes are allowed.

wisp3rwind replied:
It's definitely also redundant with the check that `_register_plugin` does.

snejus commented on beets/plugins.py:
```diff
@@ -350,57 +351,55 @@ def load_plugins(names: Sequence[str] = ()) -> None:
+
+
+def _register_plugin(cls: type[BeetsPlugin]) -> None:
+    if cls in _instances:
+        return
+
```
I doubt this check is required
```suggestion
```

wisp3rwind replied:
I think this also a left over from old code: We did use to do this repeatedly in the previous version of `find_plugins` (at least before the caching optimization was added).

If anything, it might make sense to log a warning here if a plugin is instantiated repeatedly. (Say if `load_plugins` is called repeatedly with the same plugins, which I don't think should happen.)

snejus commented on beets/plugins.py:
```diff
@@ -350,57 +351,55 @@ def load_plugins(names: Sequence[str] = ()) -> None:
+                and cls != MetadataSourcePlugin
+                and cls not in _instances
+            ):
+                classes.add(cls)
```
Do we need to store classes in this intermediate set? Can't we just `_register_plugin` here?

snejus commented on beets/plugins.py:
```diff
@@ -350,57 +351,55 @@ def load_plugins(names: Sequence[str] = ()) -> None:
-    """Returns a list of BeetsPlugin subclass instances from all
-    currently loaded beets plugins. Loads the default plugin set
-    first.
+def find_plugins() -> Iterator[BeetsPlugin]:
```
Why are we returning an iterator here instead of the list?

wisp3rwind committed:
plugins: address review

- better error handling (includes restoring some unintended changes)
- cleanup of some traces of the previous architecture of the plugin
  loader

wisp3rwind commented:
Should be good to go, but I'll rebase on https://github.com/beetbox/beets/pull/5841 once its merged.

Note that this isn't quite what the title says anymore: Logging wasn't as incomplete as I thought in the context of #5701. Thus, this PR only slightly improves logging, notably by fixing the "not found" error case (detection of which was quite hacky before, and in fact broken), and by logging the fully-qualified path to offending plugin classes.

Additionally, the PR refactors plugin loading, reducing nesting of the code, and actually instantiating plugins in `load_plugins` instead of `find_plugins` (which makes more sense conceptually, and the laziness in the former approach was irrelevant or even detrimental in practice).

snejus commented on beets/plugins.py:
```diff
@@ -355,56 +356,59 @@ def load_plugins(names: Sequence[str] = ()) -> None:
+
+
+def _register_plugin(cls: type[BeetsPlugin]) -> None:
+    if cls in _instances:
+        log.debug(
+            "repeated initialization of plugin class {}.{}",
+            cls.__module__,
+            cls.__qualname__,
+        )
```
```suggestion
```

Let's remove this check and ensure uniqueness with

```py
for name in set(names):
...
```

on line 357.

snejus reviewed (approved):
Looks good to me. I've re-run the build and added a small comment

semohr commented:
Do you feel like this was addressed? We had quite some related PRs recently.

We log the plugin name if it raises an error during import (see [here](https://github.com/beetbox/beets/blob/b06f3f6aa6e29ce3bd3f84f18ee3ff0cf5cef470/beets/plugins.py#L392)). We also test if all plugins can be loaded without an error (see [here](https://github.com/beetbox/beets/blob/b06f3f6aa6e29ce3bd3f84f18ee3ff0cf5cef470/test/test_plugins.py#L484)).

snejus commented:
> Do you feel like this was addressed? We had quite some related PRs recently.
> 
> We log the plugin name if it raises an error during import (see [here](https://github.com/beetbox/beets/blob/b06f3f6aa6e29ce3bd3f84f18ee3ff0cf5cef470/beets/plugins.py#L392)). We also test if all plugins can be loaded without an error (see [here](https://github.com/beetbox/beets/blob/b06f3f6aa6e29ce3bd3f84f18ee3ff0cf5cef470/test/test_plugins.py#L484)).

It seems like most of this has indeed been addressed now. Closing this now, @wisp3rwind if you feel like something's missing, feel free to re-open of course!
